# Thread 2: Gender Injustice

### This thread looks at the crimes against women in the past decade and concludes important pointers from it. For ex- how the rate of crime against women is affected each year, whether criminals are convicted, etc.

## EDA 

# 01. Importing libraries

In [56]:
import pandas as pd
import sys
sys.path.append("..")

import utility as u

import importlib
importlib.reload(u)

<module 'utility' from '/Users/bhavya/Documents/DS/The Committee Will Look Into It/02_Gender_Injustice/../utility.py'>

# 02. EDA on each file before using it for analysis

## a) Crime against women (2001-2022).csv

In [57]:
# this file uses wide format, we will first convert it into long format

df= u.load_data("data/01 Crime against Women (2001-2022).csv")
csv1= u.wide_to_long(df=df, id_cols="CRIME HEAD", value_name="cases", var_name="year")

In [58]:
u.eda(df=csv1)

=== DataFrame ===
Shape       : 242 rows × 3 cols
Duplicates  : 0

Column Summary:
             Dtype  Nulls  Null %  Uniques Flag
CRIME HEAD  object      0     0.0       11     
year        object      0     0.0       22     
cases        int64      0     0.0      237     

Numeric Stats:
       min       max      mean   median
cases  1.0  370616.0  45496.25  10277.5

Sample:
                                  CRIME HEAD  year  cases
0                                       Rape  2001  16075
1  Kidnapping and Abduction of Women & Girls  2001  14645
2                               Dowry Deaths  2001   6851



In [59]:
# Saving the clean file 
u.save_clean(csv1, "data/cleaned/file1.csv")

Saved | shape- (242, 3)


In [60]:
u.eda(df)

=== DataFrame ===
Shape       : 11 rows × 23 cols
Duplicates  : 0

Column Summary:
             Dtype  Nulls  Null %  Uniques Flag
CRIME HEAD  object      0     0.0       11     
2001         int64      0     0.0       11     
2002         int64      0     0.0       11     
2003         int64      0     0.0       11     
2004         int64      0     0.0       11     
2005         int64      0     0.0       11     
2006         int64      0     0.0       11     
2007         int64      0     0.0       11     
2008         int64      0     0.0       11     
2009         int64      0     0.0       11     
2010         int64      0     0.0       11     
2011         int64      0     0.0       11     
2012         int64      0     0.0       11     
2013         int64      0     0.0       11     
2014         int64      0     0.0       11     
2015         int64      0     0.0       11     
2016         int64      0     0.0       11     
2017         int64      0     0.0       11     
2018 

## Observations from csv1-
--> No nulls in the dataset.

--> Wide format converted to long: 11 crime heads × 22 years = 242 rows, 3 cols.

--> No nulls anywhere, clean dataset.

--> `year` column is object dtype after melt — needs cast to int in cleaning.

--> National level only, no state breakdown.

## b) combined_crime_against_women_full.csv

In [61]:
csv2= u.load_data("data/02 combined_crime_against_women_full.csv")

In [62]:
csv2.head(15)

,Year,State_UT,State_UT_Raw,Row_Type,Crime_Head_Standardized,Crime_Head_Raw,Is_Detailed_Subcategory,Count
0,2001,All India (Total),All India,Aggregate,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,34124.0
1,2001,All India (Total),All India,Aggregate,Cruelty by Husband or Relatives,CRUELTY BY HUSBAND OR RELATIVES,False,49170.0
2,2001,All India (Total),ALL INDIA,Aggregate,Dowry Deaths,DOWRY DEATH,False,6851.0
3,2001,All India (Total),All India,Aggregate,Immoral Traffic (Prevention) Act,IMMORAL TRAFFIC(PREVENTION)ACT,False,8796.0
4,2001,All India (Total),All India,Aggregate,Indecent Representation of Women (Prevention) Act,INDECENT REPRESENTATION OF WOMEN(PREVENTION)ACT,False,1052.0
5,2001,All India (Total),All India,Aggregate,Insult to Modesty of Women,INSULT TO THE MODESTY OF WOMEN,False,9746.0
6,2001,All India (Total),ALL INDIA K&A,Aggregate,Kidnapping & Abduction of Women,KIDNAPPING & ABDUCTION,False,14645.0
7,2001,All India (Total),All India,Aggregate,Rape,RAPE,False,16075.0
8,2001,All India (Total),TOTAL Crime against Women,Aggregate,Total Crimes against Women,TOTAL CRIMES AGAINST WOMEN,False,143795.0
9,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,19.0


In [63]:
u.eda(csv2)

=== DataFrame ===
Shape       : 6333 rows × 8 cols
Duplicates  : 0

Column Summary:
                           Dtype  Nulls  Null %  Uniques Flag
Year                       int64      0    0.00       25     
State_UT                  object      0    0.00       39     
State_UT_Raw              object      0    0.00       48     
Row_Type                  object      0    0.00        2     
Crime_Head_Standardized   object    975   15.40       15     
Crime_Head_Raw            object      0    0.00       62     
Is_Detailed_Subcategory     bool      0    0.00        2     
Count                    float64    116    1.83     2088     

Numeric Stats:
          min       max     mean  median
Year   2001.0    2025.0  2009.73  2010.0
Count     0.0  448211.0  3696.39    62.0

Sample:
   Year           State_UT State_UT_Raw   Row_Type                          Crime_Head_Standardized                                       Crime_Head_Raw  Is_Detailed_Subcategory    Count
0  2001  All India (Tot

In [64]:
csv2 = csv2[csv2["Row_Type"] == "State/UT"].reset_index(drop=True)
csv2.head(10)

,Year,State_UT,State_UT_Raw,Row_Type,Crime_Head_Standardized,Crime_Head_Raw,Is_Detailed_Subcategory,Count
0,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,19.0
1,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Cruelty by Husband or Relatives,CRUELTY BY HUSBAND OR RELATIVES,False,9.0
2,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Dowry Deaths,DOWRY DEATH,False,0.0
3,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Immoral Traffic (Prevention) Act,IMMORAL TRAFFIC(PREVENTION)ACT,False,0.0
4,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Indecent Representation of Women (Prevention) Act,INDECENT REPRESENTATION OF WOMEN(PREVENTION)ACT,False,0.0
5,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Insult to Modesty of Women,INSULT TO THE MODESTY OF WOMEN,False,1.0
6,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Kidnapping & Abduction of Women,KIDNAPPING & ABDUCTION,False,2.0
7,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Rape,RAPE,False,3.0
8,2001,Andaman & Nicobar Islands,A&N Islands,State/UT,Total Crimes against Women,TOTAL CRIMES AGAINST WOMEN,False,34.0
9,2001,Andhra Pradesh,Andhra Pradesh,State/UT,Assault on Women with Intent to Outrage Modesty,ASSAULT ON WOMEN WITH INTENT TO OUTRAGE HER MO...,False,3544.0


In [65]:
csv2["State_UT"].unique()

array(['Andaman & Nicobar Islands', 'Andhra Pradesh', 'Arunachal Pradesh',
       'Assam', 'Bihar', 'Chandigarh', 'Chhattisgarh',
       'Dadra & Nagar Haveli (pre-2020, separate UT)',
       'Daman & Diu (pre-2020, separate UT)', 'Delhi', 'Goa', 'Gujarat',
       'Haryana', 'Himachal Pradesh', 'Jammu & Kashmir', 'Jharkhand',
       'Karnataka', 'Kerala', 'Lakshadweep', 'Madhya Pradesh',
       'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland',
       'Odisha', 'Puducherry', 'Punjab', 'Rajasthan', 'Sikkim',
       'Tamil Nadu', 'Tripura', 'Uttar Pradesh', 'Uttarakhand',
       'West Bengal', 'Telangana', 'Dadra & Nagar Haveli and Daman & Diu',
       'Ladakh'], dtype=object)

Dadra & Nagar Haveli and Daman & Diu appear as separate entities pre-2020 and merged post-2020 due to the 2020 reorganisation. Will be handled in cleaning.

In [66]:
print(csv2["Year"].unique())
print(csv2["Year"].min(), "-", csv2["Year"].max())

[2001 2002 2003 2004 2005 2006 2007 2008 2009 2010 2011 2012 2013 2014
 2015 2016 2017 2018 2019 2020 2021 2022 2023 2024 2025]
2001 - 2025


The datasets looks at the crimes committed across more than a decade. 

From 2001-2025

In [67]:
csv2[csv2["Crime_Head_Standardized"].isna()][["Year","State_UT","Crime_Head_Raw"]].drop_duplicates().head(15)

,Year,State_UT,Crime_Head_Raw
4201,2014,Andaman & Nicobar Islands,1.1 - Custodial Rape
4202,2014,Andaman & Nicobar Islands,1.1.1 - Custodial_Gang Rape
4203,2014,Andaman & Nicobar Islands,1.1.2 - Custodial_Other Rape
4204,2014,Andaman & Nicobar Islands,1.2 - Rape other than Custodial
4205,2014,Andaman & Nicobar Islands,1.2.1 - Rape_Gang Rape
4206,2014,Andaman & Nicobar Islands,1.2.2 - Rape_Others
4212,2014,Andaman & Nicobar Islands,14.1 - Under ITP Section 5
4213,2014,Andaman & Nicobar Islands,14.2 - Under ITP Section 6
4214,2014,Andaman & Nicobar Islands,14.3 - Under ITP Section 7
4215,2014,Andaman & Nicobar Islands,14.4 - Under ITP Section 8


In [68]:
csv2[csv2["Count"].isna()][["Year","State_UT","Crime_Head_Standardized"]].head(20)

,Year,State_UT,Crime_Head_Standardized
5657,2015,Ladakh,Total Crimes against Women
5693,2016,Ladakh,Total Crimes against Women
5729,2017,Ladakh,Total Crimes against Women
5765,2018,Ladakh,Total Crimes against Women
5801,2019,Ladakh,Total Crimes against Women
5928,2023,Andaman & Nicobar Islands,Total Crimes against Women
5929,2023,Andhra Pradesh,Total Crimes against Women
5930,2023,Arunachal Pradesh,Total Crimes against Women
5931,2023,Assam,Total Crimes against Women
5932,2023,Bihar,Total Crimes against Women


In [69]:
csv2.groupby("Year")["Count"].apply(lambda x: x.isna().sum())

Year
2001     0
2002     0
2003     0
2004     0
2005     0
2006     0
2007     0
2008     0
2009     0
2010     0
2011     0
2012     0
2013     0
2014     0
2015     1
2016     1
2017     1
2018     1
2019     1
2020     0
2021     0
2022     0
2023    31
2024    36
2025    36
Name: Count, dtype: int64

In [70]:

# confirm Is_Detailed_Subcategory accounts for the standardized nulls
csv2[csv2["Is_Detailed_Subcategory"] == True].shape  

(900, 8)

In [71]:
csv2["Crime_Head_Standardized"].dropna().unique()

array(['Assault on Women with Intent to Outrage Modesty',
       'Cruelty by Husband or Relatives', 'Dowry Deaths',
       'Immoral Traffic (Prevention) Act',
       'Indecent Representation of Women (Prevention) Act',
       'Insult to Modesty of Women', 'Kidnapping & Abduction of Women',
       'Rape', 'Total Crimes against Women',
       'Commission of Sati (Prevention) Act', 'Dowry Prohibition Act',
       'Importation of Girls from Foreign Country',
       'Protection of Women from Domestic Violence Act',
       'Attempt to Commit Rape', 'Abetment of Suicide of Women'],
      dtype=object)

In [72]:
csv2= u.drop_totals(csv2, "Crime_Head_Raw", ["15 - Total Crimes against Women" ,"5 - Assault on Women with intent to outrage her Modesty_Total (Section 354 IPC)" ,"6 - Insult to the Modesty of Women_Total (Section 509 IPC)", "TOTAL CRIMES AGAINST WOMEN", "All India (Total)", "Total Crimes against Women"])

Dropped 563 aggregate rows from 'Crime_Head_Raw'


In [73]:
# Saving the clean file 
u.save_clean(csv2, "data/cleaned/file2.csv")

Saved | shape- (5473, 8)


## Observations from csv2
--> 6333 rows × 8 cols, covers 2001–2025. Although data is null for years 2024-25.

--> Filtered to `Row_Type == "State/UT"` to remove aggregates

--> 38 unique states/UTs

--> `Crime_Head_Standardized` has 975 nulls (15.4%) — all correspond to  `Is_Detailed_Subcategory == True` rows (granular subcategories like "1.1 - Custodial Rape", "3.2 - Kidnapping in order to Murder" etc.) that don't have a standardized label. Not data loss, just unmapped subcategories — drop or handle separately in cleaning

--> `Count` has 116 nulls (1.83%) — concentrated in 2023 (31), 2024 (36), 2025 (36) and Ladakh 2015–2019 (5). Recent years likely not fully reported yet; Ladakh pre-2020 is expected since it wasn't a separate UT then

--> 15 standardized crime heads including newer ones like `Protection of Women from Domestic Violence Act`, `Attempt to Commit Rape`, `Abetment of Suicide of Women` — not present in csv1

--> Dadra & Nagar Haveli and Daman & Diu: separate pre-2020, merged post-2020 — needs handling in cleaning

## c) NCRB_CII_2023_Table_3A.8_0.csv

In [74]:
csv3= u.load_data("data/03 NCRB_CII_2023_Table_3A.8_0.csv")
csv3.head(10)

,Sl. No.,State/UT,Cases Pending Trial from the Previous Year - ( Col. 3),Cases Sent for Trial during the year (2023) - ( Col. 4),Total Cases for Trial (Col.3+Col.4) - ( Col. 5),Cases Abated by Court - ( Col. 6),Cases Withdrawn from Prosecution - ( Col. 7),Cases Compounded or Compromised - ( Col. 8),Cases Disposed off by Plea Bargaining - ( Col. 9),Cases Quashed - ( Col. 10),...,Cases Convicted Out of Cases from Previous Year - ( Col. 13),Cases Convicted Out of Cases during the Year - ( Col. 14),Cases Convicted (Col.13+Col.14) - ( Col. 15),Cases Discharged - ( Col. 16),Cases Acquitted - ( Col. 17),Cases in which Trials were Completed (Col.15+Col.16+ Col.17) - ( Col. 18),Cases Disposed off by Courts (Col.11+Col.18) - ( Col. 19),Cases Pending Trial at End of the Year (Col.5-Col.19) - ( Col. 20),Conviction Rate (Col.15/Col.18)*100 - ( Col. 21),Pendency Percentage (Col.20/Col.5)*100 - ( Col. 22)
0,1,Andhra Pradesh,59575,21180,80755,152,6,7340,0,33,...,466,87,553,142,8855,9550,17081,63674,5.8,78.8
1,2,Arunachal Pradesh,3137,255,3392,3,0,20,1,3,...,23,1,24,8,38,70,97,3295,34.3,97.1
2,3,Assam,90075,11192,101267,227,5958,6355,1,0,...,467,265,732,36,9066,9834,22375,78892,7.4,77.9
3,4,Bihar,115796,17370,133166,0,0,338,0,0,...,1114,171,1285,65,932,2282,2620,130546,56.3,98.0
4,5,Chhattisgarh,34270,5931,40201,65,0,70,95,0,...,773,191,964,625,2777,4366,4596,35605,22.1,88.6
5,6,Goa,1811,204,2015,6,0,11,0,0,...,13,1,14,9,111,134,151,1864,10.4,92.5
6,7,Gujarat,98357,6855,105212,15,9,23,0,52,...,425,51,476,180,6531,7187,7286,97926,6.6,93.1
7,8,Haryana,41260,8987,50247,9,0,0,0,218,...,779,39,818,29,5159,6006,6233,44014,13.6,87.6
8,9,Himachal Pradesh,9370,1146,10516,3,63,25,0,111,...,135,1,136,5,707,848,1050,9466,16.0,90.0
9,10,Jharkhand,30019,6353,36372,30,1,50,1,0,...,1215,469,1684,29,3781,5494,5576,30796,30.7,84.7


In [75]:
u.eda(csv3)

=== DataFrame ===
Shape       : 39 rows × 22 cols
Duplicates  : 0

Column Summary:
                                                                                  Dtype  Nulls  Null %  Uniques Flag
Sl. No.                                                                          object      0     0.0       39     
State/UT                                                                         object      0     0.0       39     
Cases Pending Trial from the Previous Year - ( Col. 3)                            int64      0     0.0       39     
Cases Sent for Trial during the year (2023) - ( Col. 4)                           int64      0     0.0       39     
Total Cases for Trial (Col.3+Col.4) - ( Col. 5)                                   int64      0     0.0       39     
Cases Abated by Court - ( Col. 6)                                                 int64      0     0.0       26     
Cases Withdrawn from Prosecution - ( Col. 7)                                      int64      0    

In [76]:
csv3= u.drop_totals(csv3, "State/UT")
csv3.head(30)
csv3.shape

Dropped 3 aggregate rows from 'State/UT'


(36, 22)

In [77]:
csv3 = csv3.drop(columns=[
    "Sl. No.",
    "Cases Abated by Court - ( Col. 6)",
    "Cases Withdrawn from Prosecution - ( Col. 7)",
    "Cases Compounded or Compromised - ( Col. 8)",
    "Cases Disposed off by Plea Bargaining - ( Col. 9)",
    "Cases Quashed - ( Col. 10)",
    "Cases Disposed off without trial (Col.6+Col.7+Col.8+Col.9+Col.10) - ( Col. 11)",
    "Cases Stayed or Sent to Record Room - ( Col. 12)",
    "Cases Convicted Out of Cases from Previous Year - ( Col. 13)",
    "Cases Convicted Out of Cases during the Year - ( Col. 14)",
    "Cases Disposed off by Courts (Col.11+Col.18) - ( Col. 19)"
])
csv3.shape

(36, 11)

In [78]:
# Saving the clean file 
u.save_clean(csv3, "data/cleaned/file3.csv")

Saved | shape- (36, 11)


## Observations from csv3
--> **National** conviction rate is only **21.3%** — **less than 1 in 4 cases** ends in conviction.

--> Pendency is 90.8% nationally — meaning **9 out of 10 cases are still pending at year end**.

--> **UP + Maharashtra + West Bengal** alone account for the bulk of total cases — UP has **3.25 lakh cases for trial**.

--> **West Bengal** has the **highest total cases for trial (3.88 lakh)** but **conviction rate** of just **3.7%**.

--> Karnataka also having **3.2% conviction rate** — both are among the worst.

--> **Nagaland 83.3%, Mizoram 69.8%, UP 64.6%** — **highest conviction rates** but **low absolute numbers (except UP)**.

--> **Bihar pendency 98%, Arunachal 97.1%, Manipur 97.1%** — almost nothing gets resolved

--> Telangana withdrew **9,526 cases mid-trial** (Col 7+8 combined) — that's unusually high, worth flagging.

--> Ladakh has only **48 total cases, 2 convictions** — too small to draw conclusions but include for completeness.

## d) crimes-against-women-in-metros-2022.csv

In [79]:
csv4= u.load_data("data/04 crimes-against-women-in-metros-2022.csv")
csv4.head(10)

,Cities,2020,2021,2022,Population (in lakhs 2011),Rate of crime,Chargesheet rate (%)
0,Ahmedabad,1524,1554,1452,30.0,48.3,91.2
1,Bengaluru,2730,3127,3924,40.6,96.7,74.2
2,Chennai,576,874,736,43.1,17.1,81.1
3,Coimbatore,97,190,138,10.7,12.9,91.6
4,Delhi City,9782,13982,14158,75.8,186.9,72.4
5,Ghaziabad,341,591,1063,11.0,96.5,82.0
6,Hyderabad,2390,3050,3145,37.6,83.6,71.7
7,Indore,1346,1794,1809,10.4,174.3,75.9
8,Jaipur,2369,2827,3479,14.5,239.3,53.0
9,Kanpur,1056,894,2099,13.4,157.2,75.1


In [80]:
id_cols = [
    'Cities', 
    'Population (in lakhs 2011)', 
    'Rate of crime', 
    'Chargesheet rate (%)'
]

csv4 = u.wide_to_long(
    df= csv4, 
    id_cols=id_cols, 
    value_name="crime_cases", 
    var_name="year"
)
csv4.head(15)

,Cities,Population (in lakhs 2011),Rate of crime,Chargesheet rate (%),year,crime_cases
0,Ahmedabad,30.0,48.3,91.2,2020,1524
1,Bengaluru,40.6,96.7,74.2,2020,2730
2,Chennai,43.1,17.1,81.1,2020,576
3,Coimbatore,10.7,12.9,91.6,2020,97
4,Delhi City,75.8,186.9,72.4,2020,9782
5,Ghaziabad,11.0,96.5,82.0,2020,341
6,Hyderabad,37.6,83.6,71.7,2020,2390
7,Indore,10.4,174.3,75.9,2020,1346
8,Jaipur,14.5,239.3,53.0,2020,2369
9,Kanpur,13.4,157.2,75.1,2020,1056


In [81]:
csv4.shape

(60, 6)

In [82]:
u.eda(csv4)

=== DataFrame ===
Shape       : 60 rows × 6 cols
Duplicates  : 0

Column Summary:
                              Dtype  Nulls  Null %  Uniques Flag
Cities                       object      0     0.0       20     
Population (in lakhs 2011)  float64      0     0.0       20     
Rate of crime               float64      0     0.0       20     
Chargesheet rate (%)        float64      0     0.0       20     
year                         object      0     0.0        3     
crime_cases                   int64      0     0.0       60     

Numeric Stats:
                             min      max     mean   median
Population (in lakhs 2011)   9.6    540.9    54.09    17.10
Rate of crime               12.9    239.3    96.04    85.75
Chargesheet rate (%)        53.0     93.2    79.64    77.45
crime_cases                 97.0  48755.0  4250.00  1539.00

Sample:
      Cities  Population (in lakhs 2011)  Rate of crime  Chargesheet rate (%)  year  crime_cases
0  Ahmedabad                        30.0 

In [83]:
csv4['Cities'].unique()

array(['Ahmedabad', 'Bengaluru', 'Chennai', 'Coimbatore', 'Delhi City',
       'Ghaziabad', 'Hyderabad', 'Indore', 'Jaipur', 'Kanpur', 'Kochi',
       'Kolkata', 'Kozhikode', 'Lucknow', 'Mumbai', 'Nagpur', 'Patna',
       'Pune', 'Surat', 'TOTAL CITIES'], dtype=object)

In [84]:
csv4= u.drop_totals(csv4, "Cities")

Dropped 3 aggregate rows from 'Cities'


In [85]:
csv4['Cities'].unique()

array(['Ahmedabad', 'Bengaluru', 'Chennai', 'Coimbatore', 'Delhi City',
       'Ghaziabad', 'Hyderabad', 'Indore', 'Jaipur', 'Kanpur', 'Kochi',
       'Kolkata', 'Kozhikode', 'Lucknow', 'Mumbai', 'Nagpur', 'Patna',
       'Pune', 'Surat'], dtype=object)

In [86]:
csv4.shape

(57, 6)

In [87]:
# Saving the clean file 
u.save_clean(csv4, "data/cleaned/file4.csv")

Saved | shape- (57, 6)


## Observations from csv4


--> **Jaipur** has the **highest crime rate per lakh (239.3)** AND the **lowest chargesheet rate (53%)** — most crimes reported, least follow-through. worst combination in the dataset.

--> Delhi absolute numbers are staggering — **9,782 in 2020 to 14,158 in 2022**, a **44.7% jump** in **2 years**.

--> **Indore (174.3)** and **Lucknow (161.4)** have **very high crime rates** despite being **mid-sized cities** — not just a metro problem.

--> **Kolkata** is the outlier in the good direction — large city **(67.9 lakh)**, low crime rate **(27.8)**, high chargesheet rate **(86.9)**. worth noting.

--> **Chennai** is the **lowest crime rate (17.1)** among all cities despite being one of the largest (43.1 lakh).

--> **Kozhikode** and **Patna** have the **highest chargesheet rates (92.4% and 93.2%)** — cases filed actually get chargesheets. small cities doing better procedurally.

--> Total across **19 cities** went from **35,331 → 48,755** cases in just **2 years** — **38% increase**.

--> Population data is **2011 census** — crime rates are calculated on **13-year-old population figures**, so absolute rate values are underestimates for 2022.

## e) 05 Crime against Women During 2023.csv

In [88]:
csv5= u.load_data("data/05 Crime against Women During 2023.xlsx", header=None)
csv5.head(10)

,0,1,2,3,4,5,6,7,8,9,...,104,105,106,107,108,109,110,111,112,113
0,City-wise Persons Disposal Under Crime against...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,Sl. No.,53 Metropolitan Cities,PART III: A. STATUS OF PERSONS ARRESTED DURING...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,Persons to whom Notice Issued during the year ...,NaN,NaN,NaN,Persons Arrested,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,Persons whose Cases Pending Trial at End of th...,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,Persons Arrested out of Cases Registered from ...,NaN,NaN,NaN,...,NaN,NaN,Persons in whose cases Trials were completed,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,Male,Female,Transgender,Total,Male,Female,Transgender,Total,...,Transgender,Total,Male,Female,Transgender,Total,Male,Female,Transgender,Total
5,1,2,3,4,5,6,7,8,9,10,...,105,106,107,108,109,110,111,112,113,114
6,1,Agra,1277,0,0,1277,68,2,0,70,...,0,18,71,0,0,71,16312,1197,0,17509
7,2,Ahmedabad,598,408,0,1006,220,42,0,262,...,0,4647,3276,1451,0,4727,28049,9233,0,37282
8,3,Amritsar,152,28,0,180,324,45,0,369,...,0,109,148,22,0,170,1137,161,0,1298
9,4,Asansol,2294,405,0,2699,327,43,0,370,...,0,1046,1051,0,0,1051,10645,807,0,11452


In [89]:
csv5= u.clean_disposal_headers(csv5)

In [90]:
csv5= u.drop_totals(csv5, col="city", keywords=["Total Cities", "Total Cities "])

Dropped 1 aggregate rows from 'city'


In [91]:
csv5= u.drop_total_columns(csv5, id_cols=["sl_no", "cities"])

Dropped 61 aggregate/subtotal columns


In [92]:
csv5= u.wide_to_long(
    csv5,
    id_cols= ["sl_no", "city"],
    var_name= "disposable_metric",
    value_name= "person_count"
)

csv5.head(10)

,sl_no,city,disposable_metric,person_count
0,1,Agra,PART III: A. STATUS OF PERSONS ARRESTED DURING...,1277
1,2,Ahmedabad,PART III: A. STATUS OF PERSONS ARRESTED DURING...,598
2,3,Amritsar,PART III: A. STATUS OF PERSONS ARRESTED DURING...,152
3,4,Asansol,PART III: A. STATUS OF PERSONS ARRESTED DURING...,2294
4,5,Aurangabad,PART III: A. STATUS OF PERSONS ARRESTED DURING...,1200
5,6,Bengaluru,PART III: A. STATUS OF PERSONS ARRESTED DURING...,179
6,7,Bhopal,PART III: A. STATUS OF PERSONS ARRESTED DURING...,1628
7,8,Chandigarh City,PART III: A. STATUS OF PERSONS ARRESTED DURING...,22
8,9,Chennai,PART III: A. STATUS OF PERSONS ARRESTED DURING...,150
9,10,Coimbatore,PART III: A. STATUS OF PERSONS ARRESTED DURING...,124


In [93]:
csv5.shape

(2703, 4)

In [94]:
u.eda(csv5)

=== DataFrame ===
Shape       : 2703 rows × 4 cols
Duplicates  : 0

Column Summary:
                    Dtype  Nulls  Null %  Uniques Flag
sl_no              object      0     0.0       53     
city               object      0     0.0       53     
disposable_metric  object      0     0.0       51     
person_count        int64      0     0.0      537     

Numeric Stats:
              min      max   mean  median
person_count  0.0  96759.0  391.6     0.0

Sample:
  sl_no       city                                                                                                                                disposable_metric  person_count
0     1       Agra  PART III: A. STATUS OF PERSONS ARRESTED DURING THE YEAR | Persons to whom Notice Issued during the year (but not Arrested) (Sec.41 CrPC) | Male          1277
1     2  Ahmedabad  PART III: A. STATUS OF PERSONS ARRESTED DURING THE YEAR | Persons to whom Notice Issued during the year (but not Arrested) (Sec.41 CrPC) | Male           59

In [95]:
# Saving the clean file 
u.save_clean(csv5, "data/cleaned/file5.csv")

Saved | shape- (2703, 4)


## Observations from csv5


--> **53 metro cities**, **2023** data, **2703 rows** after melting (53 cities × 51 metrics).

--> **no nulls, no duplicates** — clean after parsing.

--> 114 columns in raw form — heavily nested multi-level header. needed custom parsing via `clean_disposal_headers`.

--> Metrics cover full disposal pipeline: `notice issued → arrested → chargesheeted → trial completed → pending trial` — by `gender (Male/Female/Transgender)` for each stage.

--> person_count `median` is **0** — most cities report zero for most gender-specific sub-metrics, which is expected since `Transgender` counts are **near-zero everywhere**.

--> this file complements **csv3** (state-level conviction data) — csv3 tells you state-wise outcomes, csv5 tells you city-wise person-level disposal pipeline.

--> worth noting: `disposable_metric` column contains the full pipeline label as a string — will need parsing/splitting in cleaning to extract stage and gender separately.

## f) 06 Crimes against Women in 53 Metropolitan Cities During 2023.csv

In [96]:
csv6= u.load_data("data/06 Crimes against Women in 53 Metropolitan Cities During 2023.xlsx")
csv6.head(10)

,SL,53 Metropolitan Cities,Cases Reported Under 156_3 during the year 2023,Cases Reported in Police Stations during the year 2023,Cases Reported during the year 2023
0,1.0,Agra,0,1030,1030
1,2.0,Ahmedabad,1,1359,1360
2,3.0,Amritsar,0,306,306
3,4.0,Asansol,400,295,695
4,5.0,Aurangabad,3,974,977
5,6.0,Bengaluru,68,4802,4870
6,7.0,Bhopal,0,1886,1886
7,8.0,Chandigarh City,2,369,371
8,9.0,Chennai,16,729,745
9,10.0,Coimbatore,4,240,244


In [97]:
csv6['53 Metropolitan Cities'].unique()

array(['Agra', 'Ahmedabad', 'Amritsar', 'Asansol', 'Aurangabad',
       'Bengaluru', 'Bhopal', 'Chandigarh City', 'Chennai', 'Coimbatore',
       'Delhi City', 'Dhanbad', 'Durg-Bhilainagar', 'Faridabad',
       'Ghaziabad', 'Gwalior', 'Hyderabad', 'Indore', 'Jabalpur',
       'Jaipur', 'Jamshedpur', 'Jodhpur', 'Kannur', 'Kanpur', 'Kochi',
       'Kolkata', 'Kollam', 'Kota', 'Kozhikode', 'Lucknow', 'Ludhiana',
       'Madurai', 'Malappuram', 'Meerut', 'Mumbai', 'Nagpur', 'Nasik',
       'Patna', 'Prayagraj', 'Pune', 'Raipur', 'Rajkot', 'Ranchi',
       'Srinagar', 'Surat', 'Thiruvananthapuram', 'Thrissur',
       'Tiruchirapalli', 'Vadodara', 'Varanasi', 'Vasai Virar',
       'Vijayawada', 'Vishakhapatnam', 'TOTAL CITIES'], dtype=object)

In [98]:
csv6= u.drop_totals(csv6, "53 Metropolitan Cities", keywords=['TOTAL CITIES'])

Dropped 1 aggregate rows from '53 Metropolitan Cities'


In [99]:
csv6['53 Metropolitan Cities'].unique()

array(['Agra', 'Ahmedabad', 'Amritsar', 'Asansol', 'Aurangabad',
       'Bengaluru', 'Bhopal', 'Chandigarh City', 'Chennai', 'Coimbatore',
       'Delhi City', 'Dhanbad', 'Durg-Bhilainagar', 'Faridabad',
       'Ghaziabad', 'Gwalior', 'Hyderabad', 'Indore', 'Jabalpur',
       'Jaipur', 'Jamshedpur', 'Jodhpur', 'Kannur', 'Kanpur', 'Kochi',
       'Kolkata', 'Kollam', 'Kota', 'Kozhikode', 'Lucknow', 'Ludhiana',
       'Madurai', 'Malappuram', 'Meerut', 'Mumbai', 'Nagpur', 'Nasik',
       'Patna', 'Prayagraj', 'Pune', 'Raipur', 'Rajkot', 'Ranchi',
       'Srinagar', 'Surat', 'Thiruvananthapuram', 'Thrissur',
       'Tiruchirapalli', 'Vadodara', 'Varanasi', 'Vasai Virar',
       'Vijayawada', 'Vishakhapatnam'], dtype=object)

In [100]:
csv6.head(10)

,SL,53 Metropolitan Cities,Cases Reported Under 156_3 during the year 2023,Cases Reported in Police Stations during the year 2023,Cases Reported during the year 2023
0,1.0,Agra,0,1030,1030
1,2.0,Ahmedabad,1,1359,1360
2,3.0,Amritsar,0,306,306
3,4.0,Asansol,400,295,695
4,5.0,Aurangabad,3,974,977
5,6.0,Bengaluru,68,4802,4870
6,7.0,Bhopal,0,1886,1886
7,8.0,Chandigarh City,2,369,371
8,9.0,Chennai,16,729,745
9,10.0,Coimbatore,4,240,244


In [101]:
csv6.shape

(53, 5)

In [102]:
u.eda(csv6)

=== DataFrame ===
Shape       : 53 rows × 5 cols
Duplicates  : 0

Column Summary:
                                                          Dtype  Nulls  Null %  Uniques Flag
SL                                                      float64      0     0.0       53     
53 Metropolitan Cities                                   object      0     0.0       53     
Cases Reported Under 156_3 during the year 2023           int64      0     0.0       26     
Cases Reported in Police Stations during the year 2023    int64      0     0.0       53     
Cases Reported during the year 2023                       int64      0     0.0       52     

Numeric Stats:
                                                          min      max     mean  median
SL                                                        1.0     53.0    27.00    27.0
Cases Reported Under 156_3 during the year 2023           0.0   1367.0    48.57     4.0
Cases Reported in Police Stations during the year 2023  119.0  13317.0  1396.21 

In [103]:
# Saving the clean file 
u.save_clean(csv6, "data/cleaned/file6.csv")

Saved | shape- (53, 5)


## Observations from csv6

--> **53 metro cities**, **2023** data, **53 rows × 5 columns**.

--> **no nulls, no duplicates** — fully clean dataset.

--> Tracks two main intake sources of crime against women: `Direct Police Station Reports` vs `Court Orders under Sec 156(3) CrPC`.

--> Overwhelming police reliance — **96.6% (73,999)** of total cases **(76,573)** were registered directly at **police stations**, while **court orders under 156(3)** accounted for just **3.4% (2,574)**.

--> Court-directed registration median is **4** — **26 out of 53** cities reported **0 court-directed 156(3) cases**.

--> **Jaipur** outlier in 156(3) — Jaipur alone registered **1,367 court-ordered cases**, accounting for over **53% of all 156(3) cases** reported across all 53 metro cities.

--> **Delhi City** leads total crime volume — Delhi reported **13,366 total cases** (nearly **17.5%** of the total 53-city aggregate), followed by **Mumbai (6,025)** and **Bengaluru (4,870)**.

## g) 07 Rape 2019.csv

In [104]:
csv7= u.load_data("data/07 Rape 2019.csv")
csv7.head(10)

,Category,State/UT,Punishment for Rape,Rape in Custody- Rape in Custody (Total),Rape in Custody - By Police Personnel,Rape in Custody - By Public Servant,Rape in Custody - By Member of Armed Forces,Rape in Custody - By Management/Staff of Jail/Remand Home/Place of Custody,Rape in Custody - By Management/ Staff of Hospital,Rape Other than Custodial (Total),...,Other Than Custodial Rape - Repeated Rape on same women,Punishment upto 20 years extended to life for Rape (Below 16 Yrs),Punishment for causing Death,By Separated husband,Rape/Gang Rape of Girl Below 16 Yrs (Imprisonment for life till death),By Person in Authority,Gang Rape,Rape/Gang Rape of Girl Below 12 Yrs (Imprisonment for life till death),Punishment for Repeat offenders,Rape (Total)
0,State,Andhra Pradesh,758,1,0,0,1,0,0,283,...,98,15,0,0,10,1,9,2,0,1086
1,State,Arunachal Pradesh,46,0,0,0,0,0,0,12,...,2,0,0,1,0,0,4,0,0,63
2,State,Assam,1599,11,0,0,0,11,0,104,...,0,0,1,1,1,30,25,0,0,1773
3,State,Bihar,700,0,0,0,0,0,0,16,...,0,0,0,0,0,0,14,0,0,730
4,State,Chhattisgarh,546,1,1,0,0,0,0,469,...,31,0,0,0,3,0,15,0,0,1036
5,State,Goa,54,1,0,0,0,0,1,14,...,1,1,0,0,0,0,2,0,0,72
6,State,Gujarat,429,0,0,0,0,0,0,85,...,62,0,0,0,0,0,14,0,0,528
7,State,Haryana,900,2,0,1,0,0,1,363,...,308,0,2,4,0,1,159,0,49,1480
8,State,Himachal Pradesh,302,0,0,0,0,0,0,50,...,1,2,0,0,1,0,4,0,0,359
9,State,Jammu & Kashmir,177,0,0,0,0,0,0,46,...,0,0,0,0,0,0,0,0,0,223


In [105]:
csv7['State/UT'].unique()

array(['Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar',
       'Chhattisgarh', 'Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh',
       'Jammu & Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram',
       'Nagaland', 'Odisha', 'Punjab', 'Rajasthan', 'Sikkim',
       'Tamil Nadu', 'Telangana', 'Tripura', 'Uttar Pradesh',
       'Uttarakhand', 'West Bengal', 'A & N Islands', 'Chandigarh',
       'D & N Haveli', 'Daman & Diu', 'Delhi', 'Lakshadweep',
       'Puducherry'], dtype=object)

In [106]:
csv7= u.clean_state_names(csv7, col= 'State/UT')
csv7['State/UT'].unique()

array(['Andhra Pradesh', 'Arunachal Pradesh', 'Assam', 'Bihar',
       'Chhattisgarh', 'Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh',
       'Jammu & Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram',
       'Nagaland', 'Odisha', 'Punjab', 'Rajasthan', 'Sikkim',
       'Tamil Nadu', 'Telangana', 'Tripura', 'Uttar Pradesh',
       'Uttarakhand', 'West Bengal', 'Andaman & Nicobar Islands',
       'Chandigarh', 'Dadra & Nagar Haveli', 'Daman & Diu', 'Delhi',
       'Lakshadweep', 'Puducherry'], dtype=object)

In [107]:
csv7.head(5)

,Category,State/UT,Punishment for Rape,Rape in Custody- Rape in Custody (Total),Rape in Custody - By Police Personnel,Rape in Custody - By Public Servant,Rape in Custody - By Member of Armed Forces,Rape in Custody - By Management/Staff of Jail/Remand Home/Place of Custody,Rape in Custody - By Management/ Staff of Hospital,Rape Other than Custodial (Total),...,Other Than Custodial Rape - Repeated Rape on same women,Punishment upto 20 years extended to life for Rape (Below 16 Yrs),Punishment for causing Death,By Separated husband,Rape/Gang Rape of Girl Below 16 Yrs (Imprisonment for life till death),By Person in Authority,Gang Rape,Rape/Gang Rape of Girl Below 12 Yrs (Imprisonment for life till death),Punishment for Repeat offenders,Rape (Total)
0,State,Andhra Pradesh,758,1,0,0,1,0,0,283,...,98,15,0,0,10,1,9,2,0,1086
1,State,Arunachal Pradesh,46,0,0,0,0,0,0,12,...,2,0,0,1,0,0,4,0,0,63
2,State,Assam,1599,11,0,0,0,11,0,104,...,0,0,1,1,1,30,25,0,0,1773
3,State,Bihar,700,0,0,0,0,0,0,16,...,0,0,0,0,0,0,14,0,0,730
4,State,Chhattisgarh,546,1,1,0,0,0,0,469,...,31,0,0,0,3,0,15,0,0,1036


In [108]:
csv7 = csv7.drop(columns=['Rape (Total) '])
csv7 = csv7.drop(columns=['Rape in Custody- Rape in Custody (Total)'])
csv7 = csv7.drop(columns=['Rape Other than Custodial (Total)'])
csv7.head(5)

,Category,State/UT,Punishment for Rape,Rape in Custody - By Police Personnel,Rape in Custody - By Public Servant,Rape in Custody - By Member of Armed Forces,Rape in Custody - By Management/Staff of Jail/Remand Home/Place of Custody,Rape in Custody - By Management/ Staff of Hospital,Other Than Custodial Rape- By Relative/Guardian/Teacher/person in position of trust,Other Than Custodial Rape- On Pregnant Women,...,Other Than Custodial Rape- Endanger life of Women after Rape,Other Than Custodial Rape - Repeated Rape on same women,Punishment upto 20 years extended to life for Rape (Below 16 Yrs),Punishment for causing Death,By Separated husband,Rape/Gang Rape of Girl Below 16 Yrs (Imprisonment for life till death),By Person in Authority,Gang Rape,Rape/Gang Rape of Girl Below 12 Yrs (Imprisonment for life till death),Punishment for Repeat offenders
0,State,Andhra Pradesh,758,0,0,1,0,0,17,0,...,0,98,15,0,0,10,1,9,2,0
1,State,Arunachal Pradesh,46,0,0,0,0,0,4,0,...,0,2,0,0,1,0,0,4,0,0
2,State,Assam,1599,0,0,0,11,0,31,0,...,0,0,0,1,1,1,30,25,0,0
3,State,Bihar,700,0,0,0,0,0,16,0,...,0,0,0,0,0,0,0,14,0,0
4,State,Chhattisgarh,546,1,0,0,0,0,267,2,...,3,31,0,0,0,3,0,15,0,0


In [109]:
u.eda(csv7)

=== DataFrame ===
Shape       : 36 rows × 24 cols
Duplicates  : 0

Column Summary:
                                                                                       Dtype  Nulls  Null %  Uniques Flag
Category                                                                              object      0     0.0        2     
State/UT                                                                              object      0     0.0       36     
Punishment for Rape                                                                    int64      0     0.0       32     
Rape in Custody - By Police Personnel                                                  int64      0     0.0        3     
Rape in Custody - By Public Servant                                                    int64      0     0.0        3     
Rape in Custody  - By Member of Armed Forces                                           int64      0     0.0        2     
Rape in Custody - By Management/Staff of Jail/Remand Home/Place

In [110]:
# Saving the clean file 
u.save_clean(csv7, "data/cleaned/file7.csv")

Saved | shape- (36, 24)


## Observations from csv7

--> **36 rows (States/UTs)**, **2019** data, **24 wide columns** tracking detailed situational breakdowns of rape cases.

--> **No nulls, no duplicates** — clean dataset structure.

--> Comprehensive categorization splitting offenses into `Custodial Rape` sub-types, `Other Than Custodial Rape` (relatives, victims with disabilities, minors, pregnant women), and specific `Legal/Aggravating Provisions`.

--> `Custodial offenses` are extremely rare nationally — total custodial incidents across all states are in single digits (e.g., mean for police personnel custodial rape is near **0.14**), with most states reporting 0.

--> **Family and trusted circles** dominate non-custodial risk — By `Relative`/`Guardian`/`Teacher`/`person in position of trust` has a high **mean (55.5)** and peaks up to **481** in specific states.

--> Vulnerable demographics severely impacted — sub-columns track crimes against `minors (Women Below 16 years)` - mean **21.9** and victims `incapable of giving consent` or with `mental/physical disabilities`.

--> High variance & outliers — columns like `gang rape` **(mean 53.6, max 902)** and `repeated rape on the same victim` **(mean 65.9, max 748)** show massive state-level skew.